src/Common/Epoll.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <sys/epoll.h>
+#include <vector>
+#include <boost/noncopyable.hpp>
+#include <Poco/Logger.h>
+
+namespace DB
+{
+
+using AsyncCallback = std::function<void(int, const Poco::Timespan &, const std::string &)>;
+
+class Epoll
+{
+public:
+    Epoll();
+
+    Epoll(const Epoll &) = delete;
+    Epoll & operator=(const Epoll &) = delete;
+
+    Epoll & operator=(Epoll && other);
+    Epoll(Epoll && other);
+
+    /// Add new file descriptor to epoll. If ptr set to nullptr, epoll_event.data.fd = fd,
+    /// otherwise epoll_event.data.ptr = ptr.
+    void add(int fd, void * ptr = nullptr);
+
+    /// Remove file descriptor to epoll.
+    void remove(int fd);
+
+    /// Get events from epoll. Events are written in events_out, this function returns an amount of ready events.
+    /// If blocking is false and there are no ready events,
+    /// return empty vector, otherwise wait for ready events. If blocking is true,
+    /// async_callback is given and there is no ready events, async_callback is called
+    /// with epoll file descriptor.
```

Comment:
That's unusual. Why can't we just return 0 in this case?
Looks like callback is not needed here, we can call it externally in loop.

Really, in this case I would rather write something like
```cpp
epoll_events[max_events];
size_t events_received = 0;
while (events_received < max_events)
{
    events_received += epoll.getManyReady(max_events - events_received, epoll_events + events_received, false);

    if (events_received < max_events)
        async_callback(epoll.getFileDescriptor(), ...);
}
```